### ISSUE TITLE
Race Condition Lab: renameat2 doesn't work directly

### ISSUE THREAD

**Issue by vini-fda** [seed-labs/seed-labs#64]
Hello there, professor.

I'm doing this lab using the ubuntu 20.04 VM provided as a pre-built image on https://seedsecuritylabs.org/labsetup.html, and when I try to compile the Set-UID `vulp.c` program, the GNU compiler complains that `renameat2` doesn't exist:

```C
#define _GNU_SOURCE_
  
#include <stdio.h>
#include <unistd.h>
#include <string.h>

int main()
{
    unsigned int flags = RENAME_EXCHANGE;
    char * fn = "/tmp/XYZ";
    char buffer[60];
    FILE *fp;
    /* get user input */
    scanf("%50s", buffer );

    if(!access(fn, W_OK)){
        fp = fopen(fn, "a+");
        unlink("/tmp/XYZ"); symlink("/dev/null", "/tmp/XYZ");
        unlink("/tmp/ABC"); symlink("/etc/passwd", "/tmp/ABC");
        //Does not work!
        renameat2(0, "/tmp/XYZ", 0, "/tmp/ABC", flags);
        fwrite("\n", sizeof(char), 1, fp);
        fwrite(buffer, sizeof(char), strlen(buffer), fp);
        fclose(fp);
    }
    else printf("No permission\n");
}

```

![image](https://user-images.githubusercontent.com/20410205/113346822-b6c12780-930a-11eb-866c-febd39a09567.png)

**Comment by vini-fda** [comment]
A proposed fix is to use this modified version of the program:

```C
//Modified vulp.c
//Original author: Wenliang (Kevin) Du, PhD

#define _GNU_SOURCE_


#include <sys/syscall.h>
#include <fcntl.h>
#include <stdio.h>
#include <unistd.h>
#include <string.h>

#ifndef RENAME_EXCHANGE
#define RENAME_EXCHANGE (1 << 1) /* Exchange source and dest */ // from include/uapi/linux/fs.h
#endif

int main()
{
    unsigned int flags = RENAME_EXCHANGE;
    char * fn = "/tmp/XYZ";
    char buffer[60];
    FILE *fp;
    /* get user input */
    scanf("%50s", buffer );

    if(!access(fn, W_OK)){
        fp = fopen(fn, "a+");
        unlink("/tmp/XYZ"); symlink("/dev/null", "/tmp/XYZ");
        unlink("/tmp/ABC"); symlink("/etc/passwd", "/tmp/ABC");
        //Does not work!
        //renameat2(0, "/tmp/XYZ", 0, "/tmp/ABC", flags);
        int rc = syscall(SYS_renameat2, 0, "/tmp/XYZ", 0, "/tmp/ABC", flags);
        fwrite("\n", sizeof(char), 1, fp);
        fwrite(buffer, sizeof(char), strlen(buffer), fp);
        fclose(fp);                                                                                                               1,1           Top
    }
    else printf("No permission\n");
}
```

But I'm still having trouble with running this lab properly(need to look into this further).

**Comment by kevin-w-du** [comment]
Are you doing this on the cloud or local computer? We don't have any problem doing that on the provided SEED VM (local computer). The syscall method was the one used before in Ubuntu 16.04, and I don't want to change it back.

**Comment by kevin-w-du** [comment]
You are not following the lab description. You made a mistake in `#define _GNU_SOURCE_`.

**Comment by vini-fda** [comment]
Oh pardon me, you're completely correct. It was a silly mistake.

By changing the `_GNU_SOURCE_` to the correct `_GNU_SOURCE` I was able to compile and run the lab properly.

And yes, I made other mistakes in the ordering of function calls(by inadvertently calling `fopen` before setting the symlinks, for example). After fixing these issues, the attack runs immediately(I timed it using a modified version of the attack script).

Just for documentation's sake, if anyone suffered from the same issue, [this is the new, working version of my code(spoilers)](https://gist.github.com/vini-fda/bcdf12a7da6ff18f093fa38d216728b0).

Running the attack.sh script, after doing the setup:

![image](https://user-images.githubusercontent.com/20410205/113364478-0105d100-932a-11eb-94f8-aed70401e8b6.png)

After running `sudo cat /etc/passwd`, we indeed see that the entry has been added:

![image](https://user-images.githubusercontent.com/20410205/113364577-40342200-932a-11eb-9b40-d6931feecabe.png)
